### ISSUE TITLE
The way `counter` is calculated in `PrepareProvider` can lead to unexpected interaction between seemingly isolated operations

### ISSUE THREAD

**Issue by ac10n** [ecadlabs/taquito#2680]
While writing some integration tests for possible failures, I got this kind of error:

```
"Http error response: (500) [{\"kind\":\"permanent\",\"id\":\"proto.018-Proxford.contract.manager.inconsistent_hash\",\"public_key\":\"sppk7aKNFB56X331quPePAHYuHfBR1pptMcL9zX9qWDXBRhMkUbG8SK\",\"expected_hash\":\"tz2JRU5zryWuLPffxZgUAcevDvoAPoG3vLDb\",\"provided_hash\":\"tz2TFYugWc9oh1fyEdKPzVcTAe9rUCeESKyX\"}]"
```

This is caused by having multiple operations that fail. Something like:

```typescript
import { Protocols } from "@taquito/taquito";
import { CONFIGS, isSandbox, sleep } from "./config";

CONFIGS().forEach(({ lib, rpc, setup, protocol, knownBaker }) => {
  const Tezos = lib;
  beforeAll(async (done) => {
    await setup(true);
    const setDelegateOp = await Tezos.contract.setDelegate({  
      source: await Tezos.signer.publicKeyHash(),
      delegate: knownBaker,
    });
    await setDelegateOp.confirmation();
    done();
  });
  describe(`Test staking through contract API using: ${rpc}`, () => {
    const flextesaOxford = protocol === Protocols.ProxfordS ? test : test.skip;
    beforeEach(async (done) => {
      await setup(true)
      done()
    });
    flextesaOxford('Should be able to stake', async (done) => {
      const op = await Tezos.contract.stake({
        amount: 100000,
      });
      await op.confirmation()
      expect(op.hash).toBeDefined();

      // const constants = await Tezos.rpc.getConstants();
      // await sleep(((constants.preserved_cycles + 2) * constants.blocks_per_cycle * (constants.minimal_block_delay!.toNumber())) * 1000);

      done();
    });
    flextesaOxford('Should be able to unstake', async (done) => {
      const op = await Tezos.contract.unstake({
        amount: 100000,
      });
      await op.confirmation()
      expect(op.hash).toBeDefined();

      // const constants = await Tezos.rpc.getConstants();
      // await sleep(((constants.preserved_cycles + 2) * constants.blocks_per_cycle * (constants.minimal_block_delay!.toNumber())) * 1000);

      done();
    });
    flextesaOxford('Should be able to finalizeUnstake', async (done) => {
      const op = await Tezos.contract.finalizeUnstake({});
      await op.confirmation()
      expect(op.hash).toBeDefined();
      done();
    });
    describe('Should fail with proper errors', () => {
      it.skip('Should fail with unstake amount too large, could not produce: gets proto.018-Proxford.tez.subtraction_underflow when the amount is too high', async (done) => {
        expect(async () => {
          const op = await Tezos.contract.stake({
            amount: 1000000000000,
          });
          await op.confirmation()
        }).rejects.toThrowError(/Invalid unstake request amount/);
        done();
      });
      flextesaOxford('Should fail with invalid_nonzero_transaction_amount', async (done) => {
        expect(async () => {
          const address = await Tezos.signer.publicKeyHash();
          const op = await Tezos.contract.transfer({
            to: address,
            amount: 10000,
            parameter: {
              entrypoint: "unstake",
              value: {
                int: "10000",
              },
            },
          });
          await op.confirmation()
        }).rejects.toThrowError(/\(permanent\) proto.\d{3}-\w+\.operations\.invalid_nonzero_transaction_amount/);
        done();
      });
      flextesaOxford('Should fail with invalid_unstake_request_amount', async (done) => {
        expect(async () => {
          const address = await Tezos.signer.publicKeyHash();
          const op = await Tezos.contract.transfer({
            to: address,
            amount: 0,
            parameter: {
              entrypoint: "unstake",
              value: {
                int: "-10000",
              },
            },
          });
          await op.confirmation();
        }).rejects.toThrowError(/\(permanent\) proto.\d{3}-\w+\.operations\.invalid_unstake_request_amount/);
        done();
      });
      flextesaOxford('Should fail with invalid_self_transaction_destination', async (done) => {
        expect(async () => {
          const op = await Tezos.contract.transfer({
            to: 'tz1Yju7jmmsaUiG9qQLoYv35v5pHgnWoLWbt',
            amount: 0,
            parameter: {
              entrypoint: "unstake",
              value: {
                int: "10000",
              },
            },
          });
          await op.confirmation()
        }).rejects.toThrowError(/\(permanent\) proto.\d{3}-\w+\.operations\.invalid_self_transaction_destination/);
        done();
      });
    });
  });
});
```
When running this against a testnet or flextesa, we will randomly get the error above.

I think this is caused by the implementation of `getFee`:

```
  private getFee(op: RPCOpWithFee, pkh: string, headCounter: number) {
    if (!this.#counters[pkh] || this.#counters[pkh] < headCounter) {
      this.#counters[pkh] = headCounter;
    }
    const opCounter = ++this.#counters[pkh];

    return {
      counter: `${opCounter}`,
      fee: typeof op.fee === 'undefined' ? '0' : `${op.fee}`,
      gas_limit: typeof op.gas_limit === 'undefined' ? '0' : `${op.gas_limit}`,
      storage_limit: typeof op.storage_limit === 'undefined' ? '0' : `${op.storage_limit}`,
    };
  }
```
and the hidden state: `#counters` in `PrepareProvider`. I'm still not sure, but this can possibly cause unexpected failures in Dapps.

After fixing this bug, we also need to undo any workarounds in the code that reference this issue number.

**Comment by Innkst** [comment]
We want to try and reproduce this issue with normal operations in Nairobi.
The Error "Inconsistent public key hash" exists in N, so it might happen too.

If we can't repeat this in Nairobi, this would be oxford specific.

**Comment by hui-an-yang** [comment]
tldr confirm it's not counter related and not a bug 
There is a top level `Tezos = lib` and `beforeAll` with `setup(true)` which assign Tezos with signerA then there is a nested `beforeEach` with `setup(true)` each testCases are assigning Tezos with new signerB, signerC. That’s why it’s expecting one pkh but got something else.

**Comment by hui-an-yang** [comment]
tldr confirm it's not counter related and not a bug
The reason of seeing inconsistent_hash error message is this test files has a top level `Tezos = lib` and `beforeAll` with `setup(true)` which assign Tezos with signerA then there is a nested `beforeEach` with `setup(true)` each testCases are assigning Tezos with new signerB, signerC. That’s why it’s expecting one pkh but got something else.
